scripts/update_todo_references.py

```diff
@@ -0,0 +1,248 @@
+        is_done = status_category == JIRA_DONE_STATUS_CATEGORY_ID
+        issues.append(JiraIssue(key=issue_dict["key"], done=is_done))
+
+    return issues
+
+
+def remove_todos_for_closed_issues(
+    file: Path, todos: List[Todo], issues: List[JiraIssue]
+) -> None:
+    """Removes Todos if their referenced issue was closed.
+
+    Args:
+        file: The file for which to remove the todos.
+        todos: All todos with issue reference found in this file.
+        issues: All jira issues referencing code todos.
+    """
+    open_issue_keys = {issue.key for issue in issues if not issue.done}
+    todos_to_delete = [
+        todo.issue_key
+        for todo in todos
+        if todo.issue_key not in open_issue_keys
+    ]
+
+    if todos_to_delete:
+        file_content = file.read_text()
+
+        regex = re.compile(
+            r"^[ \t]*# TODO ?\[("
+            + "|".join(todos_to_delete)
+            + r")\] ?:(.*$\n(^[ \t]*# {2}.*$\n)*)",
+            flags=re.MULTILINE
+        )
+        new_file_content = re.sub(
+            regex,
+            "",
+            file_content,
+        )
+
+        file.write_text(new_file_content)
+        print(f"Deleted {len(todos_to_delete)} todos in file '{file}'.")
+
+
+def find_todos(file: Path) -> Tuple[List[Todo], List[Todo]]:
+    """Finds referenced and unreferenced Todos in a file.
+
+    Args:
+        file: The file to search.
+
+    Returns: A tuple (unreferenced_todos, referenced_todos) for todos
+        without/with reference to jira issues.
+    """
+    file_content = file.read_text()
+
+    matches = re.findall(
+        r"^[ \t]*# TODO ?\[(LOW|MEDIUM|HIGH|[A-Z]*?-[0-9]*?)\] ?:"
+        r"(.*$\n(^[ \t]*# {2}.*$\n)*)",
+        file_content,
+        flags=re.MULTILINE,
+    )
+
+    unreferenced_todos = []
+    referenced_todos = []
+
+    for priority_or_key, description, _ in matches:
+        # remove whitespace and leading '#' from description
+        lines = description.split("\n")
+        lines = [line.strip().lstrip("#").strip() for line in lines]
+        description = " ".join(lines)
+
+        todo = Todo(filepath=str(file), description=description)
+        if priority_or_key in PRIORITIES:
+            todo.priority = priority_or_key
+            unreferenced_todos.append(todo)
+        else:
+            todo.issue_key = priority_or_key
+            referenced_todos.append(todo)
+
+    print(f"Found {len(unreferenced_todos)} unreferenced and "
+          f"{len(referenced_todos)} referenced todos in file '{file}'.")
+
+    return unreferenced_todos, referenced_todos
+
+
+def update_file_with_issue_keys(file: Path, todos: List[Todo]) -> None:
+    """Replaces TODO priorities by the issue key that the TODO is referencing.
+
+    Args:
+        file: The file in which the todos will be replaced.
+        todos: The todos to replace.
+    """
+    file_content = file.read_text()
+    todo_iterator = iter(todos)
+
+    def _replace(match):
```

Comment:
inner function :-O?

Reply:
We can keep it

Reply:
I think it actually has to be (well or we could use functools.partial, but I'd say that the inner function is cleaner here) because it needs access to the todo iterator but the function signature can only contain a single match object to work with the re.sub function

Reply:
ok